### PR TITLE
Add TapEvent headers and trailers to the tap protobuf

### LINKED ISSUE #3262: Expose arbitrary headers through tap
### What problem are you trying to solve?

Currently, when running `tap`, users are able to see a few choice details. The selection is small and has been restricted for security concerns. With tap moving to an RBAC model, these concerns are less of an issue. It is extremely helpful to see arbitrary headers as part of the request/response flow.

Tap should be expanded to show all request, response and trailer headers.

### How should the problem be solved?

There are no user facing changes for this proposal. Instead, headers should be added to the tap protobuf. As #3390 recommends a 1:1 dump of the protobuf for JSON output, the headers will automatically start showing up. This should be an unstructured list of labels similar to `EndpointMeta`.

- Include all headers, for http1 and http2 including the pseudo headers.
- Normalize header names by lowercasing them.

### Out of scope

- Proxy side filtering

### PR DESCRIPTION
**Pull request by kleimkuhler**
### Motivation

In order to expose arbitrary headers through tap, headers and trailers should be
read from the linkerd2-proxy-api `TapEvent`s and set in the public `TapEvent`s.
This change should have no user facing changes as it just prepares the events
for JSON output in linkerd/linkerd2#3390

### Solution

The public API has been updated with a headers field for
`TapEvent_Http_RequestInit_` and `TapEvent_Http_ResponseInit_`, and trailers
field for `TapEvent_Http_ResponseEnd_`.

These values are set by reading the corresponding fields off of the proxy's tap
events.

The proto changes are equivalent to the proto changes proposed in
linkerd/linkerd2-proxy-api#33

Closes #3262

Signed-off-by: Kevin Leimkuhler <kleimkuhler@icloud.com>

### CODE REVIEW

File: proto/public.proto [hunk 1]
```diff
@@ -155,6 +155,16 @@ message Scheme {
   }
 }
 
+message Headers {
+  message Header {
+    string name = 1;
+    string value = 2;
```
Comment (adleong) [inline proto/public.proto]: Probably good idea to have a comment here explaining that only one of these fields will be set, depending on if the header value is UTF-8 decodable and also explaining why we didn't just use a oneof.

Comment (kleimkuhler) [reply proto/public.proto]: Good point! I added a clarifying comment.

**Comment by kleimkuhler** [comment]
Headers and Trailers have reformatted to a flatter structure. The output is shown below:

```
{
  "source": {
    "ip": "10.244.0.1",
    "port": 51916,
    "metadata": {
      "tls": "not_provided_by_remote"
    }
  },
  "destination": {
    "ip": "10.244.0.66",
    "port": 9994,
    "metadata": {
      "control_plane_ns": "linkerd",
      "deployment": "linkerd-web",
      "namespace": "linkerd",
      "pod": "linkerd-web-54d45fc66-6dmxf",
      "pod_template_hash": "54d45fc66",
      "serviceaccount": "linkerd-web"
    }
  },
  "routeMeta": null,
  "proxyDirection": "INBOUND",
  "requestInitEvent": {
    "id": {
      "base": 0,
      "stream": 0
    },
    "method": "GET",
    "scheme": "",
    "authority": "10.244.0.66:9994",
    "path": "/ready",
    "headers": [
      {
        "name": "host",
        "valueStr": "10.244.0.66:9994"
      },
      {
        "name": "user-agent",
        "valueStr": "kube-probe/1.15"
      },
      {
        "name": "accept-encoding",
        "valueStr": "gzip"
      }
    ]
  }
}
{
  "source": {
    "ip": "10.244.0.1",
    "port": 51916,
    "metadata": {
      "tls": "not_provided_by_remote"
    }
  },
  "destination": {
    "ip": "10.244.0.66",
    "port": 9994,
    "metadata": {
      "control_plane_ns": "linkerd",
      "deployment": "linkerd-web",
      "namespace": "linkerd",
      "pod": "linkerd-web-54d45fc66-6dmxf",
      "pod_template_hash": "54d45fc66",
      "serviceaccount": "linkerd-web"
    }
  },
  "routeMeta": null,
  "proxyDirection": "INBOUND",
  "responseInitEvent": {
    "id": {
      "base": 0,
      "stream": 0
    },
    "sinceRequestInit": {
      "nanos": 354688
    },
    "httpStatus": 200,
    "headers": [
      {
        "name": "date",
        "valueStr": "Mon, 23 Sep 2019 19:56:35 GMT"
      },
      {
        "name": "content-length",
        "valueStr": "3"
      },
      {
        "name": "content-type",
        "valueStr": "text/plain; charset=utf-8"
      }
    ]
  }
}
{
  "source": {
    "ip": "10.244.0.1",
    "port": 51916,
    "metadata": {
      "tls": "not_provided_by_remote"
    }
  },
  "destination": {
    "ip": "10.244.0.66",
    "port": 9994,
    "metadata": {
      "control_plane_ns": "linkerd",
      "deployment": "linkerd-web",
      "namespace": "linkerd",
      "pod": "linkerd-web-54d45fc66-6dmxf",
      "pod_template_hash": "54d45fc66",
      "serviceaccount": "linkerd-web"
    }
  },
  "routeMeta": null,
  "proxyDirection": "INBOUND",
  "responseEndEvent": {
    "id": {
      "base": 0,
      "stream": 0
    },
    "sinceRequestInit": {
      "nanos": 382065
    },
    "sinceResponseInit": {
      "nanos": 27377
    },
    "responseBytes": 3,
    "trailers": null,
    "grpcStatusCode": 0
  }
}
```

**Comment by kleimkuhler** [comment]
The public API has changed to reflect the state of the linkerd/linkerd2-proxy-api#33, but there should be no change in behavior or output.

**Comment by ihcsim** [comment]
I'm having some difficulty building this branch:
```
✗ bin/docker-build                                                                                              
Error response from daemon: manifest for gcr.io/linkerd-io/go-deps:da0d59cc not found: manifest unknown: Failed to fetch "da0d59cc" from request "/v2/linkerd-io/go-deps/manifests/da0d59cc".

 ✗ go mod tidy                                                                                                 
go: finding github.com/linkerd/linkerd2-proxy-api v0.1.10-0.20190925194703-4b788dd3d825
go: github.com/linkerd/linkerd2-proxy-api@v0.1.10-0.20190925194703-4b788dd3d825: unknown revision 4b788dd3d825
go: error loading module requirements
```
My local `linkerd2-proxy-api` is pointing at linkerd/linkerd2-proxy-api@34ab9c848f2cadb6d3e9dc47cca5b1c36adcd59f.

Any idea?

File: cli/cmd/tap.go [hunk 2]
```diff
@@ -454,11 +478,30 @@ func getResponseEndEvent(pubEv *pb.TapEvent_Http) *responseEndEvent {
...
+		Trailers:          formatHeadersTrailers(resE.GetTrailers()),
 		GrpcStatusCode:    resE.GetEos().GetGrpcStatusCode(),
 		ResetErrorCode:    resE.GetEos().GetResetErrorCode(),
 	}
 }
 
+func formatHeadersTrailers(hs *pb.Headers) []metadata {
+	var fm []metadata
+	for _, h := range hs.GetHeaders() {
+		if _, ok := h.GetValue().(*pb.Headers_Header_ValueStr); ok {
```
Comment (ihcsim) [inline cli/cmd/tap.go]: TIOLI: You can replace these two `if`s with a `switch` type assertion:
```go
diff --git a/cli/cmd/tap.go b/cli/cmd/tap.go
index c0cffa90..59a6013e 100644
--- a/cli/cmd/tap.go
+++ b/cli/cmd/tap.go
@@ -487,16 +487,13 @@ func getResponseEndEvent(pubEv *pb.TapEvent_Http) *responseEndEvent {
 func formatHeadersTrailers(hs *pb.Headers) []metadata {
 	var fm []metadata
 	for _, h := range hs.GetHeaders() {
-		if _, ok := h.GetValue().(*pb.Headers_Header_ValueStr); ok {
+		switch h.GetValue.(type) {
+		case *pb.Headers_Header_ValueStr:
 			fht := &metadataStr{Name: h.GetName(), ValueStr: h.GetValueStr()}
 			fm = append(fm, fht)
-			continue
-		}
-
-		if _, ok := h.GetValue().(*pb.Headers_Header_ValueBin); ok {
+		case *pb.Headers_Header_ValueBin:
 			fht := &metadataBin{Name: h.GetName(), ValueBin: h.GetValueBin()}
 			fm = append(fm, fht)
-			continue
 		}
 	}
 	return fm
```

File: controller/tap/server.go [hunk 3]
```diff
@@ -357,6 +383,22 @@ func (s *GRPCTapServer) translateEvent(orig *proxy.TapEvent) *public.TapEvent {
 			}
 		}
 
+		headers := func(orig *httpPb.Headers) *public.Headers {
+			var headers []*public.Headers_Header
+			for _, header := range orig.GetHeaders() {
+				n := header.GetName()
```
Comment (ihcsim) [inline controller/tap/server.go]: Just want to point out that technically, it is possible for `orig.GetHeaders()` to return a nil. Is it possible for a `TapEvent` to have no headers?

File: controller/tap/server.go [hunk 4]
```diff
@@ -368,6 +410,7 @@ func (s *GRPCTapServer) translateEvent(orig *proxy.TapEvent) *public.TapEvent {
 							Scheme:    scheme(orig.RequestInit.GetScheme()),
 							Authority: orig.RequestInit.Authority,
 							Path:      orig.RequestInit.Path,
+							Headers:   headers(orig.RequestInit.GetHeaders()),
```
Comment (ihcsim) [inline controller/tap/server.go]: Is it possible for `GetHeaders()` to return nil? If yes, we might want to add a check in the `headers()` closure.

File: controller/tap/server.go (hunk 3 above)
Comment (kleimkuhler) [reply controller/tap/server.go]: @ihcsim Yep it is possible for a `TapEvent` to have no headers. For example it's common in `ResonseEndEvent`s

```
  "responseEndEvent": {
    "id": {
      "base": 0,
      "stream": 22
    },
    "sinceRequestInit": {
      "nanos": 8529614
    },
    "sinceResponseInit": {
      "nanos": 12929
    },
    "responseBytes": 0,
    "trailers": null,
    "grpcStatusCode": 0
  }
}
```

When `orig.GetHeaders()` returns `nil`, then it shows as above.

File: controller/tap/server.go (hunk 4 above)
Comment (kleimkuhler) [reply controller/tap/server.go]: Good call. I can add an initial `nil` check in the closure.

**Pull request merged**